lib/multiplexer/multiplexer.go

```diff
@@ -246,44 +256,145 @@ func (m *Mux) detectAndForward(conn net.Conn) {
+	// headers. After they are parsed, last pass does the actual protocol detection itself.
+	// We allow only one unsigned PROXY header from external sources, if it's enabled, and unlimited
+	// amount (within maxPROXYPasses) of signed headers from our own proxies, which take precedence.
 	var proxyLine *ProxyLine
-	for i := 0; i < 2; i++ {
+	for i := 0; i < maxPROXYPasses; i++ {
 		proto, err := detectProto(reader)
 		if err != nil {
 			return nil, trace.Wrap(err)
 		}
 
 		switch proto {
 		case ProtoProxy:
-			if !enableProxyProtocol {
+			if !m.EnableExternalProxyProtocol {
 				return nil, trace.BadParameter("proxy protocol support is disabled")
 			}
+			// We allow only one unsigned proxy line
 			if proxyLine != nil {
 				return nil, trace.BadParameter("duplicate proxy line")
 			}
 			proxyLine, err = ReadProxyLine(reader)
 			if err != nil {
 				return nil, trace.Wrap(err)
 			}
+			// repeat the cycle to detect the protocol
 		case ProtoProxyV2:
-			if !enableProxyProtocol {
-				return nil, trace.BadParameter("proxy protocol support is disabled")
+			newProxyLine, err := ReadProxyLineV2(reader)
+			if err != nil {
+				return nil, trace.Wrap(err)
+			}
+
+			// If TLVs are empty we know it can't be signed, so we don't try to verify to avoid unnecessary load
+			if m.CertAuthoritiesGetter != nil && len(newProxyLine.TLVs) > 0 {
+				hostCACerts := [][]byte{}
+				certAuthorities, err := m.CertAuthoritiesGetter.GetCertAuthorities(m.context, types.HostCA, false)
+				if err != nil {
+					return nil, trace.Wrap(err, "could not get certificate authorities")
+				}
+
+				for _, ca := range certAuthorities {
+					caCerts := services.GetTLSCerts(ca)
+					hostCACerts = append(hostCACerts, caCerts...)
+				}
+				err = newProxyLine.VerifySignature(hostCACerts, m.Clock)
+				if err != nil {
+					return nil, trace.Wrap(err, "could not verify PROXY signature")
+				}
+			}
+
+			// If proxy line is signed and successfully verified, it's always accepted as current proxy line
+			if newProxyLine.IsVerified {
+				proxyLine = newProxyLine
+				continue
 			}
+
+			// If proxy line was signed but was not successfully verified we fail
+			if m.CertAuthoritiesGetter != nil && newProxyLine.isSigned() && !newProxyLine.IsVerified {
+				return nil, trace.AccessDenied(
+					"rejected signed PROXY header, incoming connection remote/local: %q, PROXY source/destination: %q",
+					conn.RemoteAddr().String()+"->"+conn.LocalAddr().String(),
+					newProxyLine.Source.String()+"->"+newProxyLine.Destination.String())
+			}
+
+			// This is unsigned proxy line, return error if external PROXY protocol is not enabled
+			if !m.EnableExternalProxyProtocol {
+				return nil, trace.BadParameter("external proxy protocol support is disabled")
+			}
+
+			// If current proxy line was signed and verified, it takes precedence over new not signed proxy line
+			if proxyLine != nil && proxyLine.IsVerified {
```

Comment:
If we're here it means new proxy line wasn't verified (because we're continuing to the next iteration if it's verified). So if there's already exists current proxy line, which was parsed in previous iteration, and it's signed, we do not overwrite it with new not-signed proxy line, but just continue. Let me know if I'm missing something or comment needs to be amended to be more clear.

Reply:
> If we're here it means new proxy line wasn't verified

Understood.  My original question came from a concern on if we need to be referencing only the first verified proxy line, or if it's ok to replace an existing verified proxy line with another verified proxy line (thus at the end of the loop referencing the last verified proxy line).  I understand now that it does not matter, and instead we are just trying to keep from replacing a verified line with a non-verified one.

Sorry for the confusion, thank you!